src/handlers/events.rs

```diff
@@ -94,3 +126,62 @@ async fn websocket(
         }
     }
 }
+
+pub async fn console_stream(
+    ws: WebSocketUpgrade,
+    Extension(state): Extension<AppState>,
+    query: Query<WebsocketQuery>,
+) -> Result<Response, Error> {
+    let uuid = query.uuid.as_str().to_owned();
+    let users = state.users.lock().await;
+
+    let user = parse_bearer_token(query.token.as_str())
+        .and_then(|token| try_auth(&token, users.get_ref()))
+        .ok_or_else(|| Error {
+            inner: ErrorInner::PermissionDenied,
+            detail: "".to_string(),
+        })?;
+    drop(users);
+    let users = state.users.clone();
+    let event_receiver = state.event_broadcaster.subscribe();
+
+    Ok(ws
+        .on_upgrade(move |socket| console_websocket(socket, event_receiver, user.uid, uuid, users)))
+}
+
+async fn console_websocket(
+    stream: WebSocket,
+    mut event_receiver: Receiver<Event>,
+    uid: String,
+    uuid: String,
+    users: Arc<Mutex<Stateful<HashMap<String, User>>>>,
+) {
+    let (mut sender, mut _receiver) = stream.split();
+    while let Ok(event) = event_receiver.recv().await {
+        if event.instance_uuid != uuid {
+            continue;
+        }
+        match event.event_inner {
+            EventInner::InstanceOutput(_) => {
+                if can_user_view_event(
+                    &event,
+                    match users.lock().await.get_ref().get(&uid) {
```

Comment:
Can we just get the user once, instead of on every message?

Reply:
This is done for the following edge case:

1. User is authed and connected to WS
2. User is deleted before disconnecting
3. ??

In this case on the *next* event that will be send to the user, the WS connection will terminate

Reply:
Then each of the ws handler will need an event loop to notify them when a user changed, which is considerably more complex

Reply:
fine